petpals/messaging/routes.py

```diff
@@ -0,0 +1,132 @@
+from sqlalchemy import func, and_, or_, desc
+from petpals import db, socket
+from petpals.models import Messages, User
+
+router = Blueprint(
+    'message_router',
+    __name__,
+    url_prefix='/messages',
+    template_folder='templates',
+    static_folder="static",
+)
+
+
+@login_required
+@router.get('')
+def render_message_page():
+
+    # This source was very useful: https://stackoverflow.com/a/45779686
+
+    subqry = (
+        db.session.query(
+            Messages.conversation_id, func.max(Messages.time_sent).label("last_time")
+        )
+        .outerjoin(
+            User,
+            or_(
+                User.username == Messages.sender_username,
+                User.username == Messages.recipient_username,
+            ),
+        )
+        .filter(User.id == current_user.id)
+        .group_by(Messages.conversation_id)
+        .subquery()
+    )
+
+    msg_query = Messages.query.join(
+        subqry,
+        and_(
+            Messages.conversation_id == subqry.c.conversation_id,
+            Messages.time_sent == subqry.c.last_time,
+        ),
+    ).order_by(desc(Messages.time_sent))
+
+    return render_template(
+        'messages.html',
+        recent_messages=msg_query,
+    )
+
+
+@login_required
+@router.get('/new')
+def render_new_conversation_page():
+    usernames = []
+    next_conversation_id = 0
+    conversation_id_query = db.session.query(
+        func.max(Messages.conversation_id).label("highest_convo")
+    )
+
+    try:
+        next_conversation_id = int(conversation_id_query.first()['highest_convo']) + 1
+    except Exception:
+        next_conversation_id = 1
+
+    users = User.query.all()
+    messages = Messages.query.all()
+
+    for user in users:
+        if current_user.username != user.username:
+            usernames.append(user.username)
+
+    for message in messages:
+        if current_user.username == message.sender_username:
+            if message.recipient_username in usernames:
+                usernames.remove(message.recipient_username)
+        elif current_user.username == message.recipient_username:
+            if message.sender_username in usernames:
+                usernames.remove(message.sender_username)
+
+    return render_template(
+        'new_message.html',
+        usernames=sorted(usernames),
+        conversation_id=next_conversation_id,
+    )
+
+
+@login_required
+@router.post('/new/<conversation_id>')
+def create_new_conversation(conversation_id):
+    data = request.json
+    new_message = Messages()
+
+    new_message.recipient_username = data['recipient']
+    new_message.sender_username = data['sender']
+    new_message.message = data['message']
+    new_message.conversation_id = conversation_id
+
+    db.session.add(new_message)
+    db.session.commit()
+    return {'convo_id': conversation_id}
+
+
+@login_required
+@router.get("/conversation/<conversation_id>")
+def get_conversation(conversation_id):
```

Comment:
You need to verify if current_user is either sender or recipient because rn, an outside user can access the chat.